File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] firstProduct = new int[nums.length];
        int[] secondProduct = new int[nums.length];

        int product = 1;
        for (int i = 0; i < nums.length; i++) {
            firstProduct[i] = product *= nums[i];
        }
        product = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            secondProduct[i] = product *= nums[i];
        }
        int[] res = new int[nums.length];
        res[0] = secondProduct[1];
        res[nums.length - 1] = firstProduct[nums.length - 2];

        for (int i = 1; i < nums.length - 1; i++) {
            res[i] = firstProduct[i - 1] * secondProduct[i + 1];
        }

        return res;
    }
}  

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = [1] * (len(nums))

        res = 1

        for i in range(len(nums) - 2, -1, -1):
            curr = nums[i]
            
            for j in range(i, len(nums)):
                if curr < nums[j]:
                    dp[i] = max(dp[i], 1 + dp[j])       

            res = max(res, dp[i])

        return res
